perma_web/perma/tasks.py

```diff
@@ -302,586 +704,331 @@ def warn_on_exception(message="Exception in block:", exception_type=Exception):
-    target_url = link.safe_url
-
-    if link.user_deleted or link.primary_capture.status != "pending":
-        capture_job.mark_completed('deleted')
-        return
-
-    capture_job.attempt += 1
-    capture_job.save()
-
-    # Helper function to update capture_job's progress
-    def display_progress(step_count, step_description):
-        save_fields(capture_job, step_count=step_count, step_description=step_description)
-        print "%s step %s: %s" % (link.guid, step_count, step_description)
-
-    print "%s: Fetching %s" % (link.guid, target_url)
-    progress = 0
-    display_progress(progress, "Starting capture")
-
-    # suppress verbose warcprox logs
-    logging.disable(logging.INFO)
-
-    # Set up an exception we can trigger to halt capture and release all the resources involved.
-    class HaltCaptureException(Exception):
-        pass
-
-    browser = warcprox_controller = warcprox_thread = display = have_html = None
-    have_warc = False
-    thread_list = []
-    successful_favicon_urls = []
-
-    def sleep_unless_seconds_passed(seconds):
-        delta = time.time() - start_time
-        if delta < seconds:
-            wait = seconds - delta
-            print("Sleeping for {}s".format(wait))
-            time.sleep(wait)
-
-    # By domain, code to run after the target_url's page onload event.
-    # (Wrap in a lambda function to delay execution.)
-    special_domains = {
-        # Wait for splash page to auto redirect
-        "www.forbes.com": lambda: sleep_unless_seconds_passed(24),
-        "forbes.com": lambda: sleep_unless_seconds_passed(24)
-    }
-    post_load_function = special_domains.get(link.url_details.netloc)
-
-    try:
-
-        # create a request handler class that counts requests and responses
-        proxied_requests = []
-        proxied_responses = []
-        proxied_pairs = []
-        count_lock = threading.Lock()
-        class CountingRequestHandler(WarcProxyHandler):
-            def _proxy_request(self):
-
-                # make sure we don't capture anything in a banned IP range
-                if not url_in_allowed_ip_range(self.url):
-                    return
+        # Start warcprox process. Warcprox is a MITM proxy server and needs to be running
+        # before, during and after the headless browser.
+        #
+        # Start a headless browser to capture the supplied URL. Also take a screenshot if the URL is an HTML file.
+        #
+        # This whole function runs with the local dir set to a temp dir by run_in_tempdir().
+        # So we can use local paths for temp files, and they'll just disappear when the function exits.
+
+        # basic setup
+        link = capture_job.link
+        target_url = link.safe_url
+        browser = warcprox_controller = warcprox_thread = display = None
+        have_content = have_warc = have_html = False
+        thread_list = []
+        successful_favicon_urls = []
+
+        def inc_progress(step_count, step_description):
```

Comment:
Per discussion w/ Anastasia, this should probably be a method on CaptureJob instead.